crates/containerd-shim-wasmer/src/instance/instance_linux.rs

```diff
@@ -0,0 +1,263 @@
+use std::path::PathBuf;
+use std::sync::{Arc, Condvar, Mutex};
+
+use anyhow::Result;
+use containerd_shim_wasm::libcontainer_instance::LibcontainerInstance;
+use containerd_shim_wasm::sandbox::instance::ExitCode;
+use containerd_shim_wasm::sandbox::instance_utils::determine_rootdir;
+use containerd_shim_wasm::sandbox::{Error, InstanceConfig, Stdio};
+use libcontainer::container::builder::ContainerBuilder;
+use libcontainer::container::Container;
+use libcontainer::syscall::syscall::SyscallType;
+use serde::{Deserialize, Serialize};
+
+use crate::executor::WasmerExecutor;
+
+static DEFAULT_CONTAINER_ROOT_DIR: &str = "/run/containerd/wasmer";
+
+pub struct Wasi {
+    exit_code: ExitCode,
+    engine: wasmer::Cranelift,
+    stdio: Stdio,
+    bundle: String,
+    rootdir: PathBuf,
+    id: String,
+}
+
+impl LibcontainerInstance for Wasi {
+    type Engine = wasmer::Cranelift;
+
+    fn new_libcontainer(id: String, cfg: Option<&InstanceConfig<Self::Engine>>) -> Self {
+        // TODO: there are failure cases e.x. parsing cfg, loading spec, etc.
+        // thus should make `new` return `Result<Self, Error>` instead of `Self`
+        log::info!("creating new instance: {}", id);
+        let cfg = cfg.unwrap();
+        let bundle = cfg.get_bundle().unwrap_or_default();
+        let rootdir = determine_rootdir(
+            bundle.as_str(),
+            &cfg.get_namespace(),
+            DEFAULT_CONTAINER_ROOT_DIR,
+        )
+        .unwrap();
+        Wasi {
+            id,
+            exit_code: Arc::new((Mutex::new(None), Condvar::new())),
+            engine: cfg.get_engine(),
+            stdio: Stdio::init_from_cfg(cfg).expect("failed to open stdio"),
+            bundle,
+            rootdir,
+        }
+    }
+
+    fn build_container(&self) -> Result<Container, Error> {
+        let engine = self.engine.clone();
+        self.stdio.take().redirect()?;
```

Comment:
**fix:** remove.
This is causing the smoke test to time out, because it is keeping an open handle to the stdin/stdout/setderr files even after the container finished executing.

Reply:
yay, it passed smoke tests now :-)